File: maximum-depth-of-binary-tree.cpp
//Runtime: 6 ms
class Solution {
public:
    int maxd = INT_MIN;

    void solve(TreeNode* root,int d)
    {
        if(!root)
        {
            maxd = max(maxd, d);
            return ;
        }

        solve(root->left,d+1);
        solve(root->right, d+1);
    }

    int maxDepth(TreeNode* root) {
        solve(root, 0);
        return maxd;
    }
};

File: maximum-depth-of-binary-tree.scala
//Runtime: 384 ms
object Solution {
    def maxDepth(root: TreeNode): Int = {
        def solve(root: TreeNode, d: Int): Int  = root match{
            case null => 0
            case r: TreeNode if(r.left == null && r.right == null) => d
            case _ => solve(root.left, d+1) max solve(root.right, d+1)
        }
        solve(root, 1)
    }
}
